# Tools for star-typing analysis of CYP2D6

Database-driven CYP2D6 typer for full-length amplicons. Uses consensus sequences generated by [pbaa](https://github.com/PacificBiosciences/pbAA) to call star alleles using [pharmvar](https://www.pharmvar.org/gene/CYP2D6) definitions. 

## Release History
* 0.1.0 - Initial Release (05/18/2021)
* 0.1.1 - General updates (06/07/2021)
    - Update alleles to pharmvar 4.2.7.2
    - Fix vcf output (add ID, actually work)
    - Add new columns to detailed output
* 0.1.2 - Fixes/updates (06/16/2021)
    - Add option for failed consensus
    - fix dup counting/munging for trimodal alleles
* 0.1.2 - minor fixes (06/24/2021)
    - Add minLength option (default 2000bp)
    - Fix db removal of singleton sample amplicons

### Dependencies
 - [pysam](https://github.com/pysam-developers/pysam)
 - [mappy](https://pypi.org/project/mappy/)
 - [pandas](https://pandas.pydata.org/)
 - [numpy](https://numpy.org/)
 - [sqlalchemy](https://www.sqlalchemy.org/)


## Basic Usage

Call types for one or multiple samples (no vcf output):
```
$ python pbCYP2D6typer.py -r myRunID                     \
                          -s example/biosamples.csv -i   \
                          -p example/typing/multi-sample \
                          example/pbaa/*/*passed_cluster_sequences.fasta 
```
Results are in two formats
```
$ column -ts, example/typing/multi-sample_short_summary.csv
barcode         bioSample              diplotype
bc1022--bc1062  Bio_Sample_19-NA17244  *2x2/*4x2
bc1021--bc1061  Bio_Sample_20-NA17276  *2/*5
bc1011--bc1051  Bio_Sample_6-NA16688   *2/*10/*36

$ cut -d, -f1-11 example/typing/multi-sample_detailed_summary.csv | column -ts,
bioSample              barcode         cluster  numreads  seqLength  hasSpacer  coreType   type  SV   core_all         gDNA_all
Bio_Sample_19-NA17244  bc1022--bc1062  0        164       8144       0          CYP2D6_2   *2         *2;*34;*39       CYP2D6_2.001;CYP2D6_2.011;CYP2D6_2.019;CYP2D6_2.020;CYP2D6_2.029
Bio_Sample_19-NA17244  bc1022--bc1062  2        88        8673       1          CYP2D6_2   *2    DUP  *2;*34;*39       CYP2D6_2.001;CYP2D6_2.011;CYP2D6_2.019;CYP2D6_2.020;CYP2D6_2.029
Bio_Sample_19-NA17244  bc1022--bc1062  1        162       8145       0          CYP2D6_4   *4         *4;*10;*39;*74   CYP2D6_4.001;CYP2D6_4.016;CYP2D6_4.022
Bio_Sample_19-NA17244  bc1022--bc1062  3        83        8675       1          CYP2D6_4   *4    DUP  *4;*10;*39;*74   CYP2D6_4.001;CYP2D6_4.016;CYP2D6_4.022
Bio_Sample_20-NA17276  bc1021--bc1061  1        151       8144       0          CYP2D6_2   *2         *2;*34;*39       CYP2D6_2.001;CYP2D6_2.011;CYP2D6_2.019;CYP2D6_2.020;CYP2D6_2.029
Bio_Sample_20-NA17276  bc1021--bc1061  0        347       5205       0          CYP2D6_5   *5    DEL
Bio_Sample_6-NA16688   bc1011--bc1051  1        204       8145       0          CYP2D6_10  *10        *10;*39          CYP2D6_10.002
Bio_Sample_6-NA16688   bc1011--bc1051  0        226       8144       0          CYP2D6_2   *2         *2;*34;*39
Bio_Sample_6-NA16688   bc1011--bc1051  2        61        10230      1          CYP2D6_36  *36   HYB  *36;*10;*83;*39  CYP2D6_36.001;CYP2D6_36.002;CYP2D6_36.003
```
Detailed output columns:
1. bioSample (if available)
2. barcode
3. cluster
4. Number of reads in cluster
5. Sequence Length
6. CYP2D7 spacer presence
7. type
8. type
9. Structural variant class
10. All matching core sets, ordered by impact (most->least)
11. All matching star subtypes over region (gDNA)
12. All matching star subtypes over region (cDNA/exons only)

Generate per-sample VCF files with core SNV annotations (one sample at a time):
```
$ python pbCYP2D6typer.py -r myRunID                                                           \
                          -s example/biosamples.csv -i                                         \
                          -p example/typing/single_bc1011--bc1051                              \
                          --vcf                                                                \
                          --hifiSupport example/hifi/demultiplex.bc1011--bc1051.fastq          \
                          --read_info example/pbaa/bc1011--bc1051/bc1011--bc1051_read_info.txt \
                          example/pbaa/bc1011--bc1051/bc1011--bc1051_passed_cluster_sequences.fasta
```
VCF Results are phased with core SNV annotations
```
$ bcftools view -H example/typing/single_bc1011--bc1051.vcf | awk '$2>42126489 && $2<42126758'
chr22   42126611        rs1135840       C       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     1|1|1:491:30.8693,30.7325,27.5607:0,485:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126619        .       G       A       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:0,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126622        .       A       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:225,62:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126623        rs75467367      G       C       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:1,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126624        rs74478221      C       T       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:427,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126625        .       A       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:427,62:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126627        .       A       C       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:424,62:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126633        .       C       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:0,61:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126635        rs766507177     T       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:202,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126636        rs28371735      G       A       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:204,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126658        .       A       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:425,64:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126660        rs1135835       T       C       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:201,61:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126663        rs1135833       G       C       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:1,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126667        .       C       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:203,61:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
```

## Advanced Usage
Use the `--includeFail` option and pass `*failed_cluster_sequences.fasta` along with `*passed_cluster_sequences.fasta` to see failed consensus in detailed summary csv and vcf (down to `minFrac` cluster frequency).
```
$ python pbCYP2D6typer.py -r myRunID                     \
                          -s example/biosamples.csv -i   \
                          -p example/typing/multi-sample \
                          --includeFail --minFrac 0.01   \
                          example/pbaa/*/*cluster_sequences.fasta
```


Store sample records in DB for more advanced SQL access.  NOTE this is a dev feature and may contaminate outputs from the script above.  To reset DB, delete all rows from "pbaa_consensus" and "SampleVariants".
```
$ python pbCYP2D6typer.py -r myRunID                     \
                          -s example/biosamples.csv -i   \
                          -p example/typing/multi-sample \
                          --dbStore                      \
                          example/pbaa/*/*passed_cluster_sequences.fasta

```
Fun times with SQL data diving...
```
$ sqlite3 src/db/CYP2D6.db
SQLite version 3.33.0 2020-08-14 13:23:32
Enter ".help" for usage hints.
sqlite> SELECT uuid,bioSample,ID,numreads
   ...> FROM pbAA_consensus
   ...> INNER JOIN SampleVariants_annotated
   ...> USING(uuid)
   ...> WHERE Id="rs1065852"
   ...> ORDER BY bioSample;
uuid|bioSample|ID|numreads
3f056872813b7fd86d2a3c81f0b7769f|Bio_Sample_19-NA17244|rs1065852|162
d2c3139b05142d6d44961b658e2f3700|Bio_Sample_19-NA17244|rs1065852|83
18c673f4184417e62980be3b776ab8aa|Bio_Sample_6-NA16688|rs1065852|61
6de27cba1b08795a9d0f778211bd6920|Bio_Sample_6-NA16688|rs1065852|204
```
Clean up the database
```
sqlite> delete from pbAA_consensus ; delete from SampleVariants;
sqlite> .exit
```


## DISCLAIMER

THIS WEBSITE AND CONTENT AND ALL SITE-RELATED SERVICES, INCLUDING ANY DATA, ARE PROVIDED "AS IS," WITH ALL FAULTS, WITH NO REPRESENTATIONS OR WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTIES OF MERCHANTABILITY, SATISFACTORY QUALITY, NON-INFRINGEMENT OR FITNESS FOR A PARTICULAR PURPOSE. YOU ASSUME TOTAL RESPONSIBILITY AND RISK FOR YOUR USE OF THIS SITE, ALL SITE-RELATED SERVICES, AND ANY THIRD PARTY WEBSITES OR APPLICATIONS. NO ORAL OR WRITTEN INFORMATION OR ADVICE SHALL CREATE A WARRANTY OF ANY KIND. ANY REFERENCES TO SPECIFIC PRODUCTS OR SERVICES ON THE WEBSITES DO NOT CONSTITUTE OR IMPLY A RECOMMENDATION OR ENDORSEMENT BY PACIFIC BIOSCIENCES.

FOR RESEARCH USE ONLY. NOT FOR USE IN DIAGNOSTICS PROCEDURES.
