# Tools for star-typing analysis of CYP2D6

CYP2D6 star-typer for full-length HiFi amplicons. Uses consensus sequences generated by [pbaa](https://github.com/PacificBiosciences/pbAA) to call star alleles using [pharmvar](https://www.pharmvar.org/gene/CYP2D6) definitions. 

# Demo Dataset

22 Coriell samples [blah blah](https://derp)

## Release History
* 0.1.0 - Initial Release (05/18/2021)
* 0.1.1 - General updates (06/07/2021)
    - Update alleles to pharmvar 4.2.7.2
    - Fix vcf output (add ID, actually work)
    - Add new columns to detailed output
* 0.1.2 - Fixes/updates (06/16/2021)
    - Add option for failed consensus
    - fix dup counting/munging for trimodal alleles
* 0.1.3 - minor fixes (06/24/2021)
    - Add minLength option (default 2000bp)
    - Fix db removal of singleton sample amplicons
* 0.2.0 - major update (03/10/2022)
    - python3+pandas for intersections, no more processing in the db
    - snakemake workflow for one-step processing from multiplexed ccs -> star types
    - Hybrid call refinement (distinguish \*36,\*68,\*83)
    - improved diplotype munging

### Dependencies
 - [pysam](https://github.com/pysam-developers/pysam)
 - [mappy](https://pypi.org/project/mappy/)
 - [pandas](https://pandas.pydata.org/)
 - [numpy](https://numpy.org/)
 - [sqlalchemy](https://www.sqlalchemy.org/)
 - [pbaa](https://github.com/PacificBiosciences/pbAA)
 - conda
 - snakemake

## To run the pipeline
```bash
# create the base conda environment
$ conda create \
    --channel bioconda \
    --channel conda-forge \
    --prefix ./conda_env \
    python=3 snakemake=6 mamba lockfile

# activate the base conda environment
$ conda activate ./conda_env

# the workflow is in a larger repo with other tools (apps-scripts).
# clone just the CYP2D6 workflow

$ mkdir workflow && cd workflow
$ git init
$ git remote add -f origin https://github.com/PacificBiosciences/apps-scripts.git
$ git config core.sparseCheckout true
$ echo "CYP2D6tools/" >> .git/info/sparse-checkout
$ git pull origin master
$ cd ..

# to clone the whole repository

$ git clone https://github.com/jrharting/hifi-probe-capture-workflow.git workflow

# create directory for snakemake workflow

$ mkdir -p cluster_logs

# link run script for convenience

$ ln -s workflow/CYP2D6tools/run_snakemake.sh

# run the workflow for a multiplexed batch <batch_id>

$ sbatch run_snakemake.sh <batch_id> <barcode_fasta> <biosample_csv> <movie.hifi_reads.bam>
```

## Workflow Steps

1. Demultiplex samples
2. Convert bam to fastq & index
3. pbaa clustering
4. Star Typing of consensus sequences
5. Align HiFi reads to GRCh38 & paint by cluster
6. VCF generation per sample

![Snakemake workflow](https://github.com/PacificBiosciences/apps-scripts/blob/master/CYP2D6tools/images/Star-typing_wprkflow.png)

### Directory structure within basedir
 
```
.
├── batches
│   └── coriell_demo
├── cluster_logs
├── conda_env
├── run_snakemake.sh -> workflow/CYP2D6tools/run_snakemake.sh
└── workflow
     └── CYP2D6tools
```

## Worflow Outputs

Results will be generated in the directory `batches/<batch_id>`

```
$ column -ts, batches/coriell_demo/CYP2D6_star_typing_diplotype_summary.csv
bioSample  barcode         diplotype
NA02016    bc1016--bc1056  *2 x2 / *17
NA07439    bc1015--bc1055  *4 x2 / *41
NA09301    bc1014--bc1054  *1 / *2 x2
...

$ cut -d, -f1-10 batches/coriell_demo/CYP2D6_star_typing_detailed_summary.csv | column -ts,
bioSample  barcode         cluster  numreads  seqLength  hasSpacer  Type  SV   core_all                     gDNA_all
NA02016    bc1016--bc1056  2        205       8673       1          *2    DUP  *2;*34;*39                   *2.001;*2.011;*2.019;*2.020;*2.029
NA02016    bc1016--bc1056  0        406       8144       0          *2         *2;*34;*39                   *2.027
NA02016    bc1016--bc1056  1        368       8147       0          *17        *17;*2;*34;*39               *17.001
...
```



are in two formats
```
$ column -ts, example/typing/multi-sample_short_summary.csv
barcode         bioSample              diplotype
bc1022--bc1062  Bio_Sample_19-NA17244  *2x2/*4x2
bc1021--bc1061  Bio_Sample_20-NA17276  *2/*5
bc1011--bc1051  Bio_Sample_6-NA16688   *2/*10/*36

$ cut -d, -f1-11 example/typing/multi-sample_detailed_summary.csv | column -ts,
bioSample              barcode         cluster  numreads  seqLength  hasSpacer  coreType   type  SV   core_all         gDNA_all
Bio_Sample_19-NA17244  bc1022--bc1062  0        164       8144       0          CYP2D6_2   *2         *2;*34;*39       CYP2D6_2.001;CYP2D6_2.011;CYP2D6_2.019;CYP2D6_2.020;CYP2D6_2.029
Bio_Sample_19-NA17244  bc1022--bc1062  2        88        8673       1          CYP2D6_2   *2    DUP  *2;*34;*39       CYP2D6_2.001;CYP2D6_2.011;CYP2D6_2.019;CYP2D6_2.020;CYP2D6_2.029
Bio_Sample_19-NA17244  bc1022--bc1062  1        162       8145       0          CYP2D6_4   *4         *4;*10;*39;*74   CYP2D6_4.001;CYP2D6_4.016;CYP2D6_4.022
Bio_Sample_19-NA17244  bc1022--bc1062  3        83        8675       1          CYP2D6_4   *4    DUP  *4;*10;*39;*74   CYP2D6_4.001;CYP2D6_4.016;CYP2D6_4.022
Bio_Sample_20-NA17276  bc1021--bc1061  1        151       8144       0          CYP2D6_2   *2         *2;*34;*39       CYP2D6_2.001;CYP2D6_2.011;CYP2D6_2.019;CYP2D6_2.020;CYP2D6_2.029
Bio_Sample_20-NA17276  bc1021--bc1061  0        347       5205       0          CYP2D6_5   *5    DEL
Bio_Sample_6-NA16688   bc1011--bc1051  1        204       8145       0          CYP2D6_10  *10        *10;*39          CYP2D6_10.002
Bio_Sample_6-NA16688   bc1011--bc1051  0        226       8144       0          CYP2D6_2   *2         *2;*34;*39
Bio_Sample_6-NA16688   bc1011--bc1051  2        61        10230      1          CYP2D6_36  *36   HYB  *36;*10;*83;*39  CYP2D6_36.001;CYP2D6_36.002;CYP2D6_36.003
```
Detailed output columns:
1. bioSample (if available)
2. barcode
3. cluster
4. Number of reads in cluster
5. Sequence Length
6. CYP2D7 spacer presence
7. type
8. type
9. Structural variant class
10. All matching core sets, ordered by impact (most->least)
11. All matching star subtypes over region (gDNA)
12. All matching star subtypes over region (cDNA/exons only)

Generate per-sample VCF files with core SNV annotations (one sample at a time):
```
$ python pbCYP2D6typer.py -r myRunID                                                           \
                          -s example/biosamples.csv -i                                         \
                          -p example/typing/single_bc1011--bc1051                              \
                          --vcf                                                                \
                          --hifiSupport example/hifi/demultiplex.bc1011--bc1051.fastq          \
                          --read_info example/pbaa/bc1011--bc1051/bc1011--bc1051_read_info.txt \
                          example/pbaa/bc1011--bc1051/bc1011--bc1051_passed_cluster_sequences.fasta
```
VCF Results are phased with core SNV annotations
```
$ bcftools view -H example/typing/single_bc1011--bc1051.vcf | awk '$2>42126489 && $2<42126758'
chr22   42126611        rs1135840       C       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     1|1|1:491:30.8693,30.7325,27.5607:0,485:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126619        .       G       A       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:0,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126622        .       A       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:225,62:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126623        rs75467367      G       C       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:1,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126624        rs74478221      C       T       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:427,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126625        .       A       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:427,62:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126627        .       A       C       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:424,62:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126633        .       C       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:0,61:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126635        rs766507177     T       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:202,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126636        rs28371735      G       A       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:204,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126658        .       A       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:425,64:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126660        rs1135835       T       C       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:201,61:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126663        rs1135833       G       C       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:1,60:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
chr22   42126667        .       C       G       200     PASS    NS=1;AF=1       GT:DP:AQ:AD:VAF:TG:HP:DV:CH     0|0|1:491:30.8693,30.7325,27.5607:203,61:0.452,0.408,0.122:CYP2D6,CYP2D6,CYP2D6:0,1,2:0.00844131,0.0104801,0.0303279:-1,-1,0.0534979
```

## DISCLAIMER

THIS WEBSITE AND CONTENT AND ALL SITE-RELATED SERVICES, INCLUDING ANY DATA, ARE PROVIDED "AS IS," WITH ALL FAULTS, WITH NO REPRESENTATIONS OR WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTIES OF MERCHANTABILITY, SATISFACTORY QUALITY, NON-INFRINGEMENT OR FITNESS FOR A PARTICULAR PURPOSE. YOU ASSUME TOTAL RESPONSIBILITY AND RISK FOR YOUR USE OF THIS SITE, ALL SITE-RELATED SERVICES, AND ANY THIRD PARTY WEBSITES OR APPLICATIONS. NO ORAL OR WRITTEN INFORMATION OR ADVICE SHALL CREATE A WARRANTY OF ANY KIND. ANY REFERENCES TO SPECIFIC PRODUCTS OR SERVICES ON THE WEBSITES DO NOT CONSTITUTE OR IMPLY A RECOMMENDATION OR ENDORSEMENT BY PACIFIC BIOSCIENCES.

FOR RESEARCH USE ONLY. NOT FOR USE IN DIAGNOSTICS PROCEDURES.
